doc/api/advanced_input.rst

```diff
@@ -0,0 +1,91 @@
+low level in the input obtaining and implement a custom solution.
+
+Input handler classes
+---------------------
+Input handler classes are classes created to obtain a user input. If required
+:class:`InputHandler` can block application until a user input is received.
+
+This class can be instantiated everywhere in the code and used to ask for user input::
+
+    handler = InputHandler()
+    handler.get_input("Shut up and give me your input:")
+    handler.wait_for_input()
+    if handler.input_successful:
+        user_input = handler.value
+
+The :meth:`InputHandler.wait_on_input` method will block code processing until a user input is
+received. The input should be always checked if successful before processing.
+
+In the case some other work needs to be done before a user input is received then pass
+a callback to the constructor of the :class:`InputHandler` class and do not use
+the :meth:`InputHandler.wait_on_input` method. However, the callback is using
+the :ref:`event loop<event_loops_label>` so the callback won't be called until the event loop
+processing is active. If :ref:`concurrent input<concurrent_input_label>` is used then this
+callback may never be called!
+
+In the case when a user typing must be hidden then the :class:`PasswordInputHandler` class
+should be used. It shares most of the implementation with the :class:`InputHandler` but overrides
+how to obtain the code and other small nuances. For more info look on the
+:class:`PasswordInputHandler` class documentation.
+
+.. _concurrent_input_label:
+
+Concurrent input
+----------------
+
+The concurrent input is something which should be avoided. It drags an unexpected behavior to an
+application and it is hard to debug. However, there could be a time when a user input is required
+immediately even in the case that the application is already waiting for other input.
+
+By default every attempt for concurrent input will raise an exception and kill the application to
+prevent unexpected behavior. In case you want to be able to do a concurrent input then the
+:attr:`InputHandler.skip_concurrency_check` property must be set. After this property is disabled
+for the :class:`InputHandler` instance then the handler instance can make a concurrent input.
+
+The concurrent input will result in dropping all other waiting inputs so the last concurrent input
+in the queue wins a user input. The dropped waiting inputs will get a failed input signal to
+unblock :meth:`InputHandler.wait_on_input` methods.
+
+
+Creating a custom InputHandler
+------------------------------
+
+In the case the :class:`InputHandler` class or the :class:`PasswordInputHandler` class is not
+enough a developer can create its own InputHandler. The structure of input handler is based on two
+classes. First is a handler itself and second is a request object it creates.
+
+Handler class is just controller and creator of the requester and requester object is
```

Comment:
This sentence doesn't make sense to me, sorry. Maybe some commas would clear things up, but I don't know where, since I don't know what you're trying to say.